### PR TITLE
Tests+LibGL: Add a simple test case for LibGL

### PR DESCRIPTION
**Pull request by Hendiadyoin1**
This is mainly a proof of concept.
In the future we would want to regression test our implementation, and use the lagom build to cross-check against other rendering engines.

#### LibGfx: Load default font lazily
This is required when trying to use a Painter from lagom, due to
/res/font not being present
#### Tests: Add a simple LibGL render-test
At the moment we just check if we *can* render a simple triangle, we do
not yet actually test if the image is indeed the triangle we wanted.

This test also outputs the rendered image when GL_DEBUG is enabled to a
file called "picture.bmp" for manual verification.

### CODE REVIEW

File: Userland/Libraries/LibGfx/Painter.h [hunk 1]
```diff
@@ -108,7 +109,12 @@ class Painter {
     };
     void fill_path(Path const&, Color, WindingRule rule = WindingRule::Nonzero);
 
-    Font const& font() const { return *state().font; }
+    Font const& font() const
+    {
+        if (!state().font)
+            const_cast<Painter*>(this)->state().font = &FontDatabase::default_font();
```
Comment (sunverwerth) [inline Userland/Libraries/LibGfx/Painter.h:115]: Can we just `return FontDatabase::default_font();` ? No need for the `const_cast`

File: Userland/Libraries/LibGfx/Painter.cpp [hunk 2]
```diff
@@ -64,7 +64,6 @@ Painter::Painter(Gfx::Bitmap& bitmap)
     VERIFY(bitmap.physical_width() % scale == 0);
     VERIFY(bitmap.physical_height() % scale == 0);
     m_state_stack.append(State());
-    state().font = &FontDatabase::default_font();
```
Comment (sunverwerth) [inline Userland/Libraries/LibGfx/Painter.cpp:67]: `font` needs to be null initialized after these changes.

`state().font = nullptr;`

Or add a default initializer for font to the `State` struct.

File: Tests/LibGL/TestRender.cpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+#include <LibGL/GL/gl.h>
+#include <LibGL/GLContext.h>
+#include <LibGfx/BMPWriter.h>
+#include <LibGfx/Bitmap.h>
+#include <LibGfx/FontDatabase.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+#define RENDER_WIDTH 800
+#define RENDER_HEIGHT 800
```
Comment (sunverwerth) [inline Tests/LibGL/TestRender.cpp:20]: Just a general thought: We might make these dimensions smaller for the tests to run as quickly as possible. 10x10 or something should be plenty enough for simple test cases.

**Review by sunverwerth — CHANGES_REQUESTED** [review]
Great initiative! LibGL was in dire need of a test framework. My only gripes are with the `Painter` changes (see comment). I am totally unfamiliar with Lagom and Tests so someone who knows more about that should also have a look.

File: Userland/Libraries/LibGfx/Painter.h (hunk 1 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibGfx/Painter.h:115]: yeah sure

File: Tests/LibGL/TestRender.cpp (hunk 3 above)
Comment (Hendiadyoin1) [reply Tests/LibGL/TestRender.cpp:20]: should suffice for the current use-case, when we are actually testing the image, we should increase it again,
Some issues could be mere sub-pixels at that size, but we will roll with that for now then

File: Tests/LibGL/TestRender.cpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+#include <LibGfx/FontDatabase.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+#define RENDER_WIDTH 16
+#define RENDER_HEIGHT 16
+
+TEST_CASE(simple_triangle)
+{
+    auto bitmap = Gfx::Bitmap::try_create(Gfx::BitmapFormat::BGRx8888, { RENDER_WIDTH, RENDER_HEIGHT }).release_value_but_fixme_should_propagate_errors();
```
Comment (alimpfard) [inline Tests/LibGL/TestRender.cpp:24]: There's nowhere to propagate the error to, how about just
```suggestion
    auto bitmap = MUST(Gfx::Bitmap::try_create(Gfx::BitmapFormat::BGRx8888, { RENDER_WIDTH, RENDER_HEIGHT }));
```

File: Tests/LibGL/TestRender.cpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    glVertex2f(0, 1);
+    glVertex2f(-1, -1);
+    glVertex2f(1, -1);
+    glEnd();
+
+    context->present();
+
+    EXPECT_EQ(glGetError(), 0u);
+    // FIXME: Verify that the image is indeed correct
+#if GL_DEBUG
```
Comment (alimpfard) [inline Tests/LibGL/TestRender.cpp:49]: pls `if constexpr`.

**Review by alimpfard — APPROVED** [review]
LGTM if @sunverwerth is okay with the test, let's just make sure CI is happy with the -Wno-whatever changes :P

**Review by sunverwerth — APPROVED** [review]
LGTM! Let the testing begin

**Review by sunverwerth — APPROVED** [review]

**Pull request merged**